### ISSUE TITLE
rspm + docker

### ISSUE THREAD

**Issue by sgvignali** [Enchufa2/rspm#17]
Hi, thank you for this package!

Do you have any suggestions on the workflow for using `rspm` within a Docker container?

Let's say that you have a `renv` project  from a Windows host machine and you want to build a Docker container from `rocker/rstudio`.

In the `renv` webpage it is explained how to do this <https://rstudio.github.io/renv/articles/docker.html>. What would be the equivalent for `rspm`?

Sergio

**Comment by Enchufa2** [comment]
The same workflow applies if the project was initialized with `rspm::renv_init()` instead of `renv::init()`. If not (coming from Windows), then it's a bit of nightmare. You could try setting [this flag](https://rstudio.github.io/renv/reference/config.html?q=rspm#renv-config-rspm-enabled), then installing `rspm` and calling `rspm::install_sysreqs()` manually. But I found that even setting that flag, usually packages are installed from CRAN instead of RSPM.

**Comment by sgvignali** [comment]
For me the problem is not to install the pkgs from RSPM but to install the system dependencies properly. I try to provide a reproducible example, I would really appreciate your help.

1. Create an RStudio project named `rspm-test`
2. Create an R script with the following code:

```r
library(terra)
library(ggplot2)
```

3. Run `renv::init()` 
4. Create a Dockerfile with the following code:

```Dockerfile
FROM rocker/rstudio:4.2.3

# Install Ubuntu dependency
RUN apt-get update && apt-get install -y apt-file dpkg

# Install renv
ENV RENV_PATHS_CACHE /home/rstudio/.cache/R/renv
RUN Rscript -e 'install.packages("remotes")'
RUN Rscript -e 'remotes::install_github("rstudio/renv@0.17.2")'

# Restore library
WORKDIR /home/rstudio/test-rspm
COPY renv.lock .Rprofile .
RUN mkdir -p renv
COPY renv/activate.R renv/settings.json ./renv
RUN Rscript -e 'repo <- "https://packagemanager.rstudio.com/cran/latest";\
                Sys.setenv(RENV_CONFIG_REPOS_OVERRIDE = repo);\
                renv::install("rspm@0.3.0");\
                renv::restore();\
                rspm::install_sysreqs()'

# Give ownership to rstudio user
RUN chown -R rstudio:rstudio /home/rstudio
```

4. From the Terminal build the image with:

```bash
docker build --no-cache --progress=plain -t test-rspm .
```  

During the building the R packages are install from RSPM and after `rspm` tries to install the system requirements. However, there are a few warnings like the following one:

``W: Download is performed unsandboxed as root as file '/root/.local/share/R/rspm/var/lib/apt/lists/partial/archive.ubuntu.com_ubuntu_dists_jammy_InRelease' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)`

and finally and error:

```
#13 144.9 dpkg: regarding .../poppler-data_0.4.11-1_all.deb containing poppler-data, pre-dependency problem:
#13 144.9  poppler-data pre-depends on dpkg (>= 1.15.6~)
#13 144.9   dpkg is not installed.
#13 144.9
#13 144.9 dpkg: error processing archive /root/.local/share/R/rspm//var/cache/apt/archives/poppler-data_0.4.11-1_all.deb (--unpack):
#13 144.9  pre-dependency problem - not installing poppler-data
#13 144.9 Selecting previously unselected package poppler-data.
#13 144.9 Selecting previously unselected package proj-bin.
#13 144.9 Preparing to unpack .../proj-bin_8.2.1-1_amd64.deb ...
#13 144.9 Unpacking proj-bin (8.2.1-1) ...
#13 145.0 Selecting previously unselected package proj-data.
#13 145.0 Preparing to unpack .../proj-data_8.2.1-1_all.deb ...
#13 145.0 Unpacking proj-data (8.2.1-1) ...
#13 145.2 Selecting previously unselected package unixodbc-common.
#13 145.2 Preparing to unpack .../unixodbc-common_2.3.9-5_all.deb ...
#13 145.2 Unpacking unixodbc-common (2.3.9-5) ...
#13 145.2 Errors were encountered while processing:
#13 145.2  /root/.local/share/R/rspm//var/cache/apt/archives/poppler-data_0.4.11-1_all.deb
```

although `dpkg` is installed (see second command in the Dockerfile).

5. Launch the container mounting the directory with the R project:

```bash
docker run -d -p 8787:8787 -e PASSWORD=123 -v /$(pwd):/home/rstudio/test-rspm test-rspm
```

6. Go to `localhost::8787`, open the project and run `renv::restore()`. The packages look correctly installed.
7. Run `terra::rast(xmin=0, xmax=10, ymin=0, ymax=10, nrows=25, ncols=25)` 

This raises the following error:

```
Error in dyn.load(file, DLLpath = DLLpath, ...) : 
  unable to load shared object '/home/rstudio/.cache/R/renv/v5/R-4.2/x86_64-pc-linux-gnu/terra/1.7-18/5db3d690ad42a1828d3bf9e063c98587/terra/libs/terra.so':
  libproj.so.22: cannot open shared object file: No such file or directory
```

However, the file exists in the cache.

It looks like a permission problem, do you have any idea?

**Comment by Enchufa2** [comment]
Thanks for the reproducible example! Ok, so the issue with this image is that stuff is installed as root, but then RStudio runs as the rstudio user. `rspm::install_sysreqs()` must be run as the same user (i.e. rstudio), so that the libraries are installed where they can be found. BTW, you don't need to install dpkg (it's always installed; otherwise, you couldn't do anything, basically), and you can forget about the poppler error.

TL;DR, this works:

```Dockerfile
FROM rocker/rstudio:4.2.3

# Install Ubuntu dependency
RUN apt-get update && apt-get install -y apt-file

# Install renv
ENV RENV_PATHS_CACHE /home/rstudio/.cache/R/renv
RUN Rscript -e 'install.packages("remotes")'
RUN Rscript -e 'remotes::install_github("rstudio/renv@0.17.2")'

# Restore library
WORKDIR /home/rstudio/test-rspm
COPY renv.lock .Rprofile .
RUN mkdir -p renv
COPY renv/activate.R renv/settings.json ./renv
RUN Rscript -e 'repo <- "https://packagemanager.rstudio.com/cran/latest";\
                Sys.setenv(RENV_CONFIG_REPOS_OVERRIDE = repo);\
                renv::install("rspm@0.3.0");\
                renv::restore()'

# Give ownership to rstudio user
RUN chown -R rstudio:rstudio /home/rstudio

# Install system requirements as the rstudio user
RUN runuser -u rstudio -- Rscript -e 'rspm::install_sysreqs()'
```

**Comment by sgvignali** [comment]
Thank you for your response. I still get the same error for `libproj.so.22` though.

**Comment by Enchufa2** [comment]
I forgot to mention:

```
-v /$(pwd):/home/rstudio/test-rspm
```

If you do this, you are basically overwriting the installation you just made in the image, so you need to run `renv::restore()` again, and thus `rspm::install_reqs()`. Solution: do not overwrite the project folder with the mount. Or define your renv installation path in another place.

**Comment by Enchufa2** [comment]
You could set the renv library in another location, outside the project folder.

**Comment by sgvignali** [comment]
Ok, thank you. Setting the following env variable in the `Dockerfile` just before installing `renv` does the trick:

```
ENV RENV_PATHS_LIBRARY /home/rstudio/.library
```

No need to run `renv::restore()`

This is amazing, it is way easier to create the Docker image without having to worry about the system dependencies...

If you have time it would be great to have such an example in the GitHub page! I'll test it in a more complicated example to see if it works!

**Comment by Enchufa2** [comment]
Great! Please post here your final Dockerfile, and I'll update the docs with some instructions.

**Comment by sgvignali** [comment]
This is another problem, but it seems that not all dependencies for `ggplot2` are installed. In the example that I've provided, if I I create a plot with `ggplot` I get this warning:

```
Warning message:
In grSoftVersion() :
  unable to load shared object '/usr/local/lib/R/modules//R_X11.so':
  libXt.so.6: cannot open shared object file: No such file or directory
```

It seems that `libudunits2-dev` is missing. And I get another problem with `gdal` if you want I can share a script to recreate the problem, but I can test it only in Docker. Probably it is better to create another issue.

**Comment by Enchufa2** [comment]
Yes, please, thank you.

**Comment by Enchufa2** [comment]
Shorter Dockerfile, with the fix in #17:

```Dockerfile
FROM rocker/rstudio:4.2.3

# Install Ubuntu dependency
RUN apt-get update && apt-get install -y apt-file

# Install renv
ENV RENV_PATHS_CACHE /home/rstudio/.cache/R/renv
ENV RENV_PATHS_LIBRARY /home/rstudio/.library
ENV RENV_CONFIG_REPOS_OVERRIDE https://packagemanager.rstudio.com/cran/latest

# Restore library
WORKDIR /home/rstudio/test-rspm
COPY renv.lock .Rprofile .
RUN mkdir -p renv
COPY renv/activate.R renv/settings.json ./renv
RUN Rscript -e 'renv::restore()'

# Give ownership to rstudio user
RUN chown -R rstudio:rstudio /home/rstudio

# Install system requirements as the rstudio user
RUN runuser -u rstudio -- \
  Rscript -e 'renv::install("rspm@0.3.1"); rspm::install_sysreqs()'
ENV PROJ_LIB /home/rstudio/.local/share/R/rspm/usr/share/proj
```

**Comment by Enchufa2** [comment]
I'm adding a root mode in the `feature/root` branch for improved compatibility with Docker workflows. You can test it with this Dockerfile:

```Dockerfile
FROM rocker/rstudio:4.2.3

# Install Ubuntu dependency
RUN apt-get update && apt-get install -y apt-file

# Install renv
ENV RENV_PATHS_CACHE /home/rstudio/.cache/R/renv
ENV RENV_PATHS_LIBRARY /home/rstudio/.library
ENV RENV_CONFIG_REPOS_OVERRIDE https://packagemanager.rstudio.com/cran/latest

# Restore library
WORKDIR /home/rstudio/test-rspm
COPY renv.lock .Rprofile .
RUN mkdir -p renv
COPY renv/activate.R renv/settings.json ./renv
RUN Rscript -e 'renv::restore()'
RUN Rscript -e 'renv::install("Enchufa2/rspm@feature/root"); rspm::install_sysreqs()'

# Give ownership to rstudio user
RUN chown -R rstudio:rstudio /home/rstudio
```

**Comment by sgvignali** [comment]
Thanks for checking this, I'm in holidays, I will test it after Easter

**Comment by sgvignali** [comment]
It works perfectly! I've simplified the `Dockerfile` by copying only the `renv.lock` file:

```Dockerfile
FROM rocker/rstudio:4.2.3

# Install Ubuntu dependency
RUN apt-get update && apt-get install -y apt-file

# Install renv
ENV RENV_PATHS_CACHE /home/rstudio/.cache/R/renv
ENV RENV_PATHS_LIBRARY /home/rstudio/.library
ENV RENV_CONFIG_REPOS_OVERRIDE https://packagemanager.rstudio.com/cran/latest

# Restore library
WORKDIR /home
COPY renv.lock .
RUN Rscript -e 'renv::restore()'
RUN Rscript -e 'renv::install("Enchufa2/rspm@feature/root"); rspm::install_sysreqs()'

# Give ownership to rstudio user
RUN chown -R rstudio:rstudio /home/rstudio
```

Notice that the `renv.lock` file is copied in the `home` directory so it won't be overwritten by mounting the project directory.

Are you planning to release this change on CRAN?

**Comment by Enchufa2** [comment]
Great, thanks for testing. Yes, I'll merge this and release soon.

**Comment by Enchufa2** [comment]
On CRAN now.

**Comment by sgvignali** [comment]
Great, thank you!
